### ISSUE TITLE
Smartparens and Common Lisp

### ISSUE THREAD

**Issue by fpopineau** [Fuco1/smartparens#303]
Hi,
Don't know if it has been reported and it may be an unfrequent case, but I happened to work on some CL code involving stuff like:
(foo |rdf|:|someClass|)
If you try to 'push' right paren (with Ctrl- right arrow) through some symbol like |rdf|:|someClass|, the paren jumps after '|rdf|', then after ':' and last after '|someClass|'.
And you end up with spaces around the ':'
I hope it is not too hard to take into account that such things are symbols too.
Because SP is much nicer than paredit to me.
Thanks for this very useful package.

**Comment by wobh** [comment]
Sorry in advance for the length, but this is one of those issues that made choosing between Paredit and Smartparens for Common Lisp development really tough.

Paredit does this:

<pre>
()foo:bar
 ^ slurp-forward
(foo:bar)
</pre>

which seems right for parens, but it also does:

<pre>
||foo:bar
 ^ slurp-forward
|foo:bar|
</pre>

which seems wrong.

Also wrong is this:

<pre>
()f|oo|:b|a|r
 ^ slurp-forward
(f)|oo|:b|a|r
 ^ slurp-forward
(f|oo|):b|a|r
 ^ slurp-forward
(f|oo|:b)|a|r
 ^ slurp-forward
(f|oo|:b|a|)r
 ^ slurp-forward
(f|oo|:b|a|r)
</pre>

The right thing is to do what it did in the first example.

An unescaped `|` isn't a symbol character, but a token for the CL reader, and so is `:` for that matter. I think the challenge here is supporting things like `F|oo|:B|a|R` because that's how to make selected parts of the symbol literal, while the rest is left to the reader. Why would anyone do that? So that a mixed case symbol can be read consistently when `readtable-case` is other values. Paraphrasing from the Hyperspec:

```lisp
(let ((*readtable* (copy-readtable *readtable* nil)))
  (format t "READTABLE-CASE | Input          | Symbol-name~
           ~%---------------|----------------|------------~
           ~%")
  (dolist (readtable-case '(:upcase :downcase :preserve :invert))
    (setf (readtable-case *readtable*) readtable-case)
    (dolist (input '("f|oo|\\:b|A|r" "F|Oo|\\:b|a|R"))
      (format t "~&`:~A`~15T| `~A`~30T| `~A`"
	      (string-upcase readtable-case)
	      input
	      (symbol-name (read-from-string input))))))
```
READTABLE-CASE | Input          | Symbol-name
---------------|----------------|------------
`:UPCASE`      | `f|oo|\:b|A|r` | `Foo:BAR`
`:UPCASE`      | `F|Oo|\:b|a|R` | `FOo:BaR`
`:DOWNCASE`    | `f|oo|\:b|A|r` | `foo:bAr`
`:DOWNCASE`    | `F|Oo|\:b|a|R` | `fOo:bar`
`:PRESERVE`    | `f|oo|\:b|A|r` | `foo:bAr`
`:PRESERVE`    | `F|Oo|\:b|a|R` | `FOo:baR`
`:INVERT`      | `f|oo|\:b|A|r` | `Foo:BAR`
`:INVERT`      | `F|Oo|\:b|a|R` | `FOo:baR`

The example is extreme but the issue is fairly mundane. The OP did exactly this in the `|foo|:|bar|` example to keep from treating `:` as part of the symbol-name for the lisp reader, because package resolution is done by the reader.

String-quote seems like the right syntax class for Emacs, but in the context of the lisp reader, characters on either side of it are still part of the symbol name being read. Inserting spaces around `|` isn't right, because `|` does not separate symbols, it's doing something unrelated. If it's escaped it's a regular symbol character and not a delimiter. I think slurping or barfing in `||` should gulp up one character at a time.

Here's an exception, though, a case of Paredit doing something slightly dubious:

<pre>
||(foo:bar)
 ^ slurp-forward
|(foo:bar)|
</pre>

Now, this is probably the user's intention, and as a default behavior that would be okay. However, if we have to complicate this with syntax class exceptions, I'd love the ability to configure those so that I could make it do `|(|foo:bar)` instead, or perhaps signal an "invalid context" error.

For `:` it makes sense to me to treat it as a word separator rather than a separator of symbols. `foo:bar` is, for all practical purposes, one symbol. In all the lisps I'm aware of, only whitespace separates symbols, and, in CL only if it's not between two unescaped `|`.

I don't know if this can work in Smartparens. If if it could, I'd switch.

###### References
- http://l1sp.org/cl/2.2
- http://l1sp.org/cl/2.3
- http://l1sp.org/cl/23.1.2.1

**Comment by Fuco1** [comment]
Thanks for the writeup.  I'll reopen this, but this can probably only get scheduled to release 1.9 or 2.1, which comes gods know when.

I will re-read it again when I'm less tired and maybe try to come up with some ideas for someone else to hack on.

**Comment by Fuco1** [comment]
The issue from the OP was fixed in 1.10 so I'm going to close this issue.

As for the examples put out by @wobh SP seems not to break the spacing so the code stays "correct".  The user still has to invoke the slurp multiple times though.

If we want to work on this let's open a new issue with a specific goal.
